src/nni_evaluation_engine.hpp

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019-2022 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+//
+// The NNI Evaluation Engine is an interface between the NNI Engine, providing different
+// methods for scoring NNIs.
+
+#pragma once
+
+#include "gp_dag.hpp"
+#include "graft_dag.hpp"
+#include "nni_engine_key_index.hpp"
+
+using NNIDoubleMap = std::map<NNIOperation, double>;
+
+class NNIEngine;
+class TPEngine;
+class GPEngine;
+
+class NNIEvaluationEngine {
+ public:
+  using KeyIndex = NNIEngineKeyIndex;
+  using KeyIndexPairArray = NNIEngineKeyIndexPairArray;
+  using KeyIndexMap = NNIEngineKeyIndexMap;
+  using KeyIndexMapPair = NNIEngineKeyIndexMapPair;
+
+  explicit NNIEvaluationEngine(NNIEngine &nni_engine);
+
+  virtual ~NNIEvaluationEngine() {}
+
+  // Initialize Evaluation Engine.
+  virtual void Init() { Failwith("Pure virtual function call."); }
```

Comment:
@ognian- is this the usual thing to do to avoid making a virtual call?

Reply:
Yes, that's useable when we want to be able to instantiate NNIEvaluationEngine directly, but keep some of the functions unimplemented.